### DISCUSSION TITLE
Customised toString() function

### DISCUSSION THREAD

**Discussion by medha-kb** [kobylynskyi/graphql-java-codegen#1415] [Q&A]
Hello,

I am using the `graphql-java-codegen` plugin for generating Kotlin classes from a GraphQL schema. I am looking for a way to modify the toString method in the generated classes to include custom logging.

data class Book(val id: String, val title: String) {
    override fun toString(): String {
        val stringRepresentation = "Book(id=$id, title=$title)"
        // logging 
        log.info("Book toString called: $stringRepresentation")
        return stringRepresentation
    }
}

 Is there a way  that allows this level of customization in the generated Kotlin classes via custom templates or some other configuration option? Or would you recommend an alternative approach to customize the behavior of methods in the generated classes?

Any guidance or suggestions on this would be greatly appreciated.

Thank you!

**Comment by kobylynskyi** [comment]
Thanks for the interest in the library.
Currently we don't support customizing generated methods.

**Reply by jxnu-liguobin** [reply]
This may require customizing the entire template?

**Reply by medha-kb** [reply]
Following up on my previous discussion, I've attempted to set up a custom template and updated my pom.xml accordingly. However, the generated Kotlin data classes do not reflect the intended changes in the toString() method. My current objective is to append a product code to the toString() output of these classes to facilitate easier parsing of logs		

Here is the toString() method I am trying to implement in my custom_type.ftl 


		override fun toString(): String {
			val joiner = StringJoiner(", ", "{ ", " }")
			joiner.add("productCode: XYZ123") // Adding custom product code.
			<#if fields?has_content>
				<#list fields as field>
					<#if field.type?ends_with("?")>
						if (${field.name} != null) {
							// existing conditional serialization logic...
						}
					
			return joiner.toString()
		}
	
pom.xml 

<configuration>
    <customTemplates>
        <type>/templates/custom_type.ftl</type>
    </customTemplates>
    
</configuration>


I've double-checked the path, ensured the syntax is correct, and performed a mvn clean install to no avail. Could you please point me toward any documentation on custom template implementation or provide insight into what might be missing in my setup?

I appreciate your help!

**Reply by kobylynskyi** [reply]
@medha-kb please share your plugin configuration options, along with a path to the template. Thanks

**Reply by medha-kb** [reply]
Thank you for your response. Due to confidentiality reasons, I'm unable to share complete configuration details. However, I've added a new section for the custom template under the <configuration> tag, similar to:

\```
```xml
<customTemplates>
    <type>path/to/custom_type.ftl</type>
</customTemplates>


This path is adjusted to fit our project's directory structure. Also, could you advise on the recommended version of the plugin for optimal support of custom templates?

**Reply by kobylynskyi** [reply]
First of, try changing the type to uppercase: TYPE.
Also look at the following related issues:

https://github.com/kobylynskyi/graphql-java-codegen/issues/1244

https://github.com/kobylynskyi/graphql-java-codegen/issues/1158

Maybe you might need to supply a customTepmplatesRoot parameter.
Hope that helps.

**Reply by medha-kb** [reply]
Thank you, I will try the solutions suggested.

**Reply by medha-kb** [reply]
The custom template works now with the addition of customTemplatesRoot parameter. 

However, we have a dependency on another service that is still using an older version of the plugin, specifically v4.1.x
I am interested in knowing if  these two versions are compatible with each other. The primary concern is to ensure that the project can continue to work smoothly while using the latest version of the plugin, even though one of our dependencies is on an older version.
